torchaudio/functional/filtering.py

```diff
@@ -877,12 +892,18 @@ def lfilter(
 
     input_signal_windows.div_(a_coeffs[0])
     a_coeffs_flipped.div_(a_coeffs[0])
-    for i_sample, o0 in enumerate(input_signal_windows.t()):
-        windowed_output_signal = padded_output_waveform[
-            :, i_sample:i_sample + n_order
-        ]
-        o0.addmv_(windowed_output_signal, a_coeffs_flipped, alpha=-1)
-        padded_output_waveform[:, i_sample + n_order - 1] = o0
+
+    if input_signal_windows.device == torch.device('cpu') and\
+       a_coeffs_flipped.device == torch.device('cpu') and\
+       padded_output_waveform.device == torch.device('cpu'):
+        lfilter_core_loop(input_signal_windows, a_coeffs_flipped, padded_output_waveform)
+    else:
+        for i_sample, o0 in enumerate(input_signal_windows.t()):
+            windowed_output_signal = padded_output_waveform[
+                :, i_sample:i_sample + n_order
+            ]
+            o0.addmv_(windowed_output_signal, a_coeffs_flipped, alpha=-1)
+            padded_output_waveform[:, i_sample + n_order - 1] = o0
```

Comment:
Is the content of `else` clause same as the `lfilter_core_loop` function defined at line 15? If so, can we reuse it?

Reply:
and if we cannot reuse it, can we leave a comment on why?

Reply:
> and if we cannot reuse it, can we leave a comment on why?

 Unfortunately, we cannot reuse it in else clause because depending on module availability the lfilter_core_loop may point to C++ operator in which case we cannot use it if the tensors are not on CPU. The reason we need to follow this complex logic is combination of following reasons:
- C++ operator is not available on Windows
- Try loop cannot be inside function as TS would complain
- lfilter_core_loop cannot be initialized with None (in which case we could have simply added this non-nullability check on if clause) because TS cannot infer type and would complain during compilation.
- An alternative solution would be to initialize lfilter_core_loop with any 'type' allowed by TS and add check on if statement that lfilter_core_loop is not same as what it was initialized with. In this case we could have avoided code repetition. But initialization with arbitrary type might have created confusion so I avoided it :). 


> Is the content of `else` clause same as the `lfilter_core_loop` function defined at line 15? If so, can we reuse it?

Yes, the content of else clause is same as lfilter_core_loop function defined at line 15.

Reply:
Did you try something like this?

```python
def lfilter_core_generic_loop(
        input_signal_windows: Tensor,
        a_coeffs_flipped: Tensor,
        padded_output_waveform: Tensor,
):
    n_order = a_coeffs_flipped.size(0)
    for i_sample, o0 in enumerate(input_signal_windows.t()):
        windowed_output_signal = padded_output_waveform[
            :, i_sample:i_sample + n_order
        ]
        o0.addmv_(windowed_output_signal, a_coeffs_flipped, alpha=-1)
        padded_output_waveform[:, i_sample + n_order - 1] = o0

try:
    lfilter_core_cpu_loop = torch.ops.torchaudio._lfilter_core_loop
except RuntimeException:
    lfilter_core_cpu_loop = lfilter_core_generic_loop
```

then inside of `lfilter`

```python
def lfilter(...)
    ...
    if input_signal_windows.device == torch.device('cpu') and\
       a_coeffs_flipped.device == torch.device('cpu') and\
       padded_output_waveform.device == torch.device('cpu'):
        lfilter_core_cpu_loop(input_signal_windows, a_coeffs_flipped, padded_output_waveform)
    else:
        lfilter_core_generic_loop(input_signal_windows, a_coeffs_flipped, padded_output_waveform)

Reply:
Thank you @mthrok for the suggestion. This makes sense. Let me commit the new changes accordingly.